pkg/governance/provider.go

```diff
@@ -0,0 +1,170 @@
+ "issuanceDate": "2020-03-16T22:37:26.544Z",
+ "issuer": {
+   "id": "did:example:oakek12as93mas91220dapop092",
+   "name": "Governance"
+ },
+ "credentialSubject": {
+   "name": "trustbloc",
+   "version": "1.0",
+   "logo": "https://example.com/logo",
+   "description": "Trustbloc Governs.",
+   "docs_uri": "https://example.com/docs",
+   "data_uri": "https://example.com/data.json",
+   "topics": ["banking"],
+   "jurisdictions": ["ca"],
+   "geos": ["Canadian"],
+   "roles": ["accreditor"],
+   "privileges": [
+       {"name": "accredit", "uri": "https://example.com/accredit"}
+   ],
+  "duties": [
+       {"name": "safe-accredit", "uri": "https://example.com/responsible-accredit"}
+    ],
+   "define": [
+       {"name": "DID", "id": "%s"}
+   ]
+ }
+}`
+
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// issueCredentialRequest request for issuing credential.
+type issueCredentialRequest struct {
+	Credential json.RawMessage `json:"credential,omitempty"`
+}
+
+// Provider provide governance operation.
+type Provider struct {
+	governanceVCSUrl string
+	httpClient       httpClient
+	store            storage.Store
+	requestTokens    map[string]string
+}
+
+// New return new provider for governance provider.
+func New(governanceVCSUrl string, tlsConfig *tls.Config, s storage.Provider,
+	requestTokens map[string]string) (*Provider, error) {
+	err := s.CreateStore(storeName)
+	if err != nil && !errors.Is(err, storage.ErrDuplicateStore) {
+		return nil, fmt.Errorf("failed to create store: %w", err)
+	}
+
+	store, err := s.OpenStore(storeName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open store : %w", err)
+	}
+
+	return &Provider{governanceVCSUrl: governanceVCSUrl,
+		httpClient: &http.Client{Transport: &http.Transport{TLSClientConfig: tlsConfig}}, store: store,
+		requestTokens: requestTokens}, nil
+}
+
+// IssueCredential issue credential.
+func (p *Provider) IssueCredential(didID string) ([]byte, error) {
+	_, err := p.GetCredential()
+	if err != nil {
+		if errors.Is(err, storage.ErrValueNotFound) {
+			return p.issueCredential(didID)
+		}
+
+		return nil, err
+	}
+
+	return nil, fmt.Errorf("governance vc already issued")
+}
+
+func (p *Provider) issueCredential(didID string) ([]byte, error) {
+	req := &issueCredentialRequest{
+		Credential: []byte(fmt.Sprintf(governanceJSONLdDoc, didID)),
+	}
+
+	reqBytes, err := json.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+
+	endpointURL := fmt.Sprintf(issueCredentialURLFormat, p.governanceVCSUrl, governanceProfileName)
+
+	httpReq, err := http.NewRequest(http.MethodPost, endpointURL, bytes.NewBuffer(reqBytes))
+
+	if err != nil {
+		return nil, err
+	}
+
+	return sendHTTPRequest(httpReq, p.httpClient, http.StatusCreated, p.requestTokens[vcsGovernanceRequestTokenName])
+}
+
+// GetCredential get governance credential.
+func (p *Provider) GetCredential() ([]byte, error) {
+	return p.store.Get(governanceVCKey)
```

Comment:
Looks to be a single goverance vc for a deployment. If we are creating at profile/tenant level shouldn't we have multiple governance VCs ?

Reply:
done